flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/AbstractMetricGroup.java

```diff
@@ -83,9 +88,25 @@
 
 	// ------------------------------------------------------------------------
 
-	public AbstractMetricGroup(MetricRegistry registry, String[] scope) {
+	public AbstractMetricGroup(MetricRegistry registry, String[] scope, A parent) {
 		this.registry = checkNotNull(registry);
 		this.scopeComponents = checkNotNull(scope);
+		this.parent = parent;
+	}
+
+	public Map<String, String> getAllVariables() {
+		if (variables == null) { // avoid synchronization for common case
```

Comment:
Yeah, `isEmpty()` wouldn't work. 

I checked my implementation and for JMX it would not be the data-processing thread that calls ´getAllVariables()`, so this would in fact be alright. For our current non-JMX reporters this would be done in the respective reporter thread.